addons/binding/org.openhab.binding.rfxcom/src/main/java/org/openhab/binding/rfxcom/internal/connector/RFXComStreamReader.java

```diff
@@ -37,89 +36,50 @@ public RFXComStreamReader(RFXComBaseConnector connector, InputStream in) {
-                    logger.trace("Received data (len={}): {}", len, DatatypeConverter.printHexBinary(logData));
-
-                    for (int i = 0; i < len; i++) {
-
-                        if (index > dataBufferMaxLen) {
-                            // too many bytes received, try to find new
-                            // start
-                            startFound = false;
-                        }
-
-                        if (!startFound && tmpData[i] > 0) {
-
-                            startFound = true;
-                            index = 0;
-                            dataBuffer[index++] = tmpData[i];
-                            msgLen = tmpData[i] + 1;
-
-                        } else if (startFound) {
-
-                            dataBuffer[index++] = tmpData[i];
-
-                            if (index == msgLen) {
-                                sendMessage(dataBuffer, msgLen);
-
-                                // find new start
-                                startFound = false;
-                            }
+        final int MAX_READ_TIMEOUTS = 4;
+        byte[] buf = new byte[Byte.MAX_VALUE];
+
+        // The stream has (or SHOULD have) a read timeout set. Taking a
+        // read timeout (read returns 0) between packets gives us a chance
+        // to check if we've  been interrupted. Read interrupts during a
+        // packet are ignored but if too many occur we take it as meaning
+        // the RFXCOM has become missing presumed dead.
+        while (!Thread.interrupted()) {
+            try {
+                // First byte tells us how long the packet is
+                int bytesRead = in.read(buf, 0, 1);
+                int packetLength = buf[0];
+
+                if (bytesRead > 0 && packetLength > 0) {
+                    // Now read the rest of the packet
+                    int bufferIndex = 1;
+                    int readTimeoutCount = 1;
+
+                    while (bufferIndex <= packetLength) {
+                        int bytesRemaining = packetLength - bufferIndex + 1;
+                        bytesRead = in.read(buf, bufferIndex, bytesRemaining);
+
+                        if (bytesRead > 0) {
+                            bufferIndex += bytesRead;
+                        } else if (readTimeoutCount++ == MAX_READ_TIMEOUTS) {
+                            throw new IOException("Timeout during packet read");
                         }
                     }
-                } else {
-                    try {
-                        Thread.sleep(100);
-                    } catch (InterruptedException ignore) {
-                    }
+
+                    connector.sendMsgToListeners(Arrays.copyOfRange(buf, 0, packetLength + 1));
+                }
+            } catch (IOException ioe) {
+                connector.sendErrorToListeners(ioe.getMessage());
+                try {
+                    Thread.sleep(100);
+                } catch (InterruptedException ie) {
```

Comment:
If we receive this exception, the interrupted flag was cleared and we need to reset the flag:

http://www.yegor256.com/2015/10/20/interrupted-exception.html
https://www.ibm.com/developerworks/library/j-jtp05236/